node/txnSyncConn.go

```diff
@@ -0,0 +1,236 @@
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+// Package node is the Algorand node itself, with functions exposed to the frontend
+package node
+
+import (
+	"context"
+	"time"
+
+	"github.com/algorand/go-algorand/data"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/network"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/txnsync"
+	"github.com/algorand/go-algorand/util/timers"
+)
+
+// transactionSyncNodeConnector implements the txnsync.NodeConnector interface, allowing the
+// transaction sync communicate with the node and it's child objects.
+type transactionSyncNodeConnector struct {
+	node           *AlgorandFullNode
+	eventsCh       chan txnsync.Event
+	clock          timers.WallClock
+	messageHandler txnsync.IncomingMessageHandler
+	txHandler      data.SolicitedAsyncTxHandler
+	openStateCh    chan struct{}
+}
+
+func makeTransactionSyncNodeConnector(node *AlgorandFullNode) transactionSyncNodeConnector {
+	return transactionSyncNodeConnector{
+		node:        node,
+		eventsCh:    make(chan txnsync.Event, 1),
+		clock:       timers.MakeMonotonicClock(time.Now()),
+		txHandler:   node.txHandler.SolicitedAsyncTxHandler(),
+		openStateCh: make(chan struct{}),
+	}
+}
+
+func (tsnc *transactionSyncNodeConnector) Events() <-chan txnsync.Event {
+	return tsnc.eventsCh
+}
+
+// GetCurrentRoundSettings is called when the txsync is starting up, proving
+// round information.
+func (tsnc *transactionSyncNodeConnector) GetCurrentRoundSettings() txnsync.RoundSettings {
+	round := tsnc.node.ledger.Latest()
+	return txnsync.RoundSettings{
+		Round:             round,
+		FetchTransactions: tsnc.node.config.ForceFetchTransactions || tsnc.node.accountManager.HasLiveKeys(round, round),
+	}
+}
+
+// NotifyMonitor is used for testing purposes only, and can remain(almost) empty on production code.
+func (tsnc *transactionSyncNodeConnector) NotifyMonitor() chan struct{} {
+	return tsnc.openStateCh
+}
+
+func (tsnc *transactionSyncNodeConnector) Random(rng uint64) uint64 {
+	return tsnc.node.Uint64() % rng
+}
+
+func (tsnc *transactionSyncNodeConnector) Clock() timers.WallClock {
+	return tsnc.clock
+}
+
+func (tsnc *transactionSyncNodeConnector) GetPeer(networkPeer interface{}) txnsync.PeerInfo {
+	unicastPeer := networkPeer.(network.UnicastPeer)
+	if unicastPeer == nil {
+		return txnsync.PeerInfo{}
+	}
+
+	peerData := tsnc.node.net.GetPeerData(networkPeer, "txsync")
+	if peerData == nil {
```

Comment:
The if check is needed in the current casting usage.
It's true that if you were to write
out, _ := me.(*other)
then it would be fine, even when me==nil.

Reply:
Why is it needed though? I think behavior should be the same if we removed the if statement.

Reply:
no, it wouldn't.. but I can rewrite this as 
```golang
func (tsnc *transactionSyncNodeConnector) GetPeer(networkPeer interface{}) txnsync.PeerInfo {
	unicastPeer := networkPeer.(network.UnicastPeer)
	if unicastPeer == nil {
		return txnsync.PeerInfo{}
	}

	peerData := tsnc.node.net.GetPeerData(networkPeer, txnsyncPeerDataKey)
        out := txnsync.PeerInfo{
			IsOutgoing:  unicastPeer.IsOutgoing(),
			NetworkPeer: unicastPeer,
		}
        out.TxnSyncPeer, _ = peerData.(*txnsync.Peer)
        return out
}
```

if you feel that this makes it any cleaner.

Reply:
What would be different if we skipped `if`? Still not getting it. :)

Reply:
if you cast from nil and you have a single output variable, it panics.
if you cast from nil and you have a two output variables, the first one receive the nil, and the second one get "false".

Reply:
Got it, thanks! The current version is fine.